### ISSUE TITLE
Fleet controller consuming high cpu

### ISSUE THREAD

**Issue by StrongMonkey** [rancher/fleet#606]
SURE-3613

**Issue description**:
fleet controller is consuming a large amount of CPU, possibly querying the Kubernetes API server at a high rate, which is impacting other k8s services that consume the API.

**Business impact**:
This might be impacting the customers ability to deploy a certain chart or affect the K8S api servers availability.

**Troubleshooting steps**:
deploy a hurt chart? I believe in this case it could be due to a chart trying to interpret `deletionGracePeriodSeconds`

**Repro steps**:
deploy a hurt chart? I believe in this case it could be due to a chart trying to interpret `deletionGracePeriodSeconds`

**Additional notes**:
Seeing a lot of this in the fleet controller
```
gitjobs(map[objectset.rio.cattle.io/hash:88bd6abcf757f7fb2377854899cb72a44cc5482a]) 1.004423631s"
time="2021-11-05T00:07:12Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:7734899c5a839e11a1f4e2361e91d27084de0146]) 1.005166593s"
time="2021-11-05T00:07:12Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:3aa974cf2fbc6cee42ac2569146e5c253243e1b4]) 1.003825323s"
time="2021-11-05T00:07:12Z" level=info msg="rate limited gitjobs(map[objectset.rio.cattle.io/hash:96a01aeda7a8eaf214e354b37837e0c3fa77073e]) 1.00407412s"
time="2021-11-05T00:07:12Z" level=info msg="rate limited
```

From `top`:
```
top - 00:13:54 up 30 days, 23:51, 0 users, load average: 6.87, 6.41, 5.18
Tasks: 203 total, 2 running, 201 sleeping, 0 stopped, 0 zombie
%Cpu(s): 57.6 us, 15.3 sy, 0.2 ni, 26.3 id, 0.0 wa, 0.0 hi, 0.7 si, 0.0 st
MiB Mem : 7705.7 total, 163.6 free, 4436.1 used, 3105.9 buff/cache
MiB Swap: 0.0 total, 0.0 free, 0.0 used. 4137.6 avail Mem

PID USER PR NI VIRT RES SHR S %CPU %MEM TIME+ COMMAND
3735043 ubuntu 20 0 913240 232788 42364 S 66.1 3.0 37:16.76 fleetcontroller
3733659 root 20 0 926080 194860 81784 S 28.2 2.5 11:04.69 containerd
3733647 root 20 0 3103152 2.3g 90524 S 27.6 30.0 14:38.21 k3s-server
```

It’s hammering our environment pretty hard and starting to impact performance. Don’t think it started recently. Last 30 days have a lot of CPU/load
image.png

Also looks like the gitjob container crashed. Last 10 lines in log:
```
$ kubectl logs -n fleet-system gitjob-568c57cfb9-fwj64 --previous --tail=10
time="2021-11-04T16:10:51Z" level=error msg="error syncing 'fleet-default/indexing-automation-staging': handler sync-repo: failed to delete fleet-default/indexing-automation-staging-87197 batch/v1, Kind=Job for sync-repo fleet-default/indexing-automation-staging: jobs.batch \"indexing-automation-staging-87197\" not found, requeuing"
time="2021-11-04T16:10:51Z" level=error msg="error syncing 'fleet-default/indexing-automation-staging': handler sync-repo: failed to delete fleet-default/indexing-automation-staging-87197 batch/v1, Kind=Job for sync-repo fleet-default/indexing-automation-staging: jobs.batch \"indexing-automation-staging-87197\" not found, requeuing"
time="2021-11-04T16:10:51Z" level=error msg="error syncing 'fleet-default/indexing-automation-staging': handler sync-repo: failed to delete fleet-default/indexing-automation-staging-87197 batch/v1, Kind=Job for sync-repo fleet-default/indexing-automation-staging: jobs.batch \"indexing-automation-staging-87197\" not found, requeuing"
time="2021-11-04T16:10:51Z" level=error msg="error syncing 'fleet-default/indexing-automation-staging': handler sync-repo: failed to delete fleet-default/indexing-automation-staging-87197 batch/v1, Kind=Job for sync-repo fleet-default/indexing-automation-staging: jobs.batch \"indexing-automation-staging-87197\" not found, requeuing"
```
Amongst the above errors, this appears to be repeated in the fleet-controller logs roughly every minute:
```
time="2021-11-05T00:29:41Z" level=error msg="error syncing 'fleet-default/docassemble-staging-docassemble': handler bundle: error while running post render on files: json: cannot unmarshal array into Go value of type map[string]interface {}, requeuing"
```
The related GitRepo based on the name, has a Ready status failure which may be historical. The error above appears to be from the conversion to a helm chart, perhaps the latest commit has invalid yaml? Would these retries trigger fleet to consume this amount of CPU?

that might be a historical commit where the manifest was valid, but was updating an immutable field
```
status:
commit: 0812f6684008507222792f3a10ad55d25307071d
conditions:

lastUpdateTime: "2021-09-28T14:45:29Z"
message: 'ErrApplied(1) [Bundle docassemble-staging-docassemble: cannot patch
"docassemble-staging-docassemble-redis" with kind Deployment: Deployment.apps
"docassemble-staging-docassemble-redis" is invalid: metadata.deletionGracePeriodSeconds:
Invalid value: 120: field is immutable]; deployment.apps docassemble-staging/docassemble-staging-docassemble-minio
[progressing] Pending termination: 1; deployment.apps docassemble-staging/docassemble-staging-docassemble-postgres
[progressing] Pending termination: 1; deployment.apps docassemble-staging/docassemble-staging-docassemble-redis
modified {"metadata":{"deletionGracePeriodSeconds":120}}; service.v1 docassemble-staging/docassemble-staging-docassemble-minio-svc
extra'
status: "False"
type: Ready
```

**Comment by sadhugit** [comment]
Hi Team, Any update on this issue. Like ETA.

**Comment by ritheshgm** [comment]
+1, We are waiting for the fix. In our case we don't see any error/warning in logs as shown in attachment. 

<img width="1086" alt="image" src="https://user-images.githubusercontent.com/50571179/175869151-872fc7f0-9f4d-48bc-a615-64db702819c5.png">

**Comment by moio** [comment]
Does anybody know how to reliably reproduce this bug in a test environment?

**Comment by janeczku** [comment]
Issue exists on v0.3.11 and v0.5.0-rc2
SURE-4967

**Comment by kkaempf** [comment]
Waiting for @moio

**Comment by bmdepesa** [comment]
@kkaempf If there is nothing specific to test here for v2.7.2 can we move it out of the milestone?

**Comment by moio** [comment]
I am still waiting on feedback and keeping relevant people pinged (every two weeks). As this is not really working I will propose a deadline next week.

**Comment by moio** [comment]
Made another round of checking and, AFAICS, in all known cases symptoms have either been resolved via fixes merged in recent versions or users stopped responding.

Thus, I am going ahead and close this issue now.

If any user still experiences abnormally high CPU usage from `fleet-controller` using Rancher >= 2.7.2/2.6.11 or Fleet standalone >= 0.5.3 I would kindly ask to open a new issue, as this one has accumulated a long history making it hard to navigate.

Please include:
 - Rancher/Fleet versions
 - Number of clusters, GitRepos, bundles
 - A fresh log tarball ([instructions](https://www.suse.com/support/kb/doc/?id=000020191))
 - Data/graphs from monitoring

Feel free to CC me directly (@moio).

Thanks everybody for the cooperation!

**Comment by sadhugit** [comment]
One of customer(who upgrade to Rancher v2.7.2) still notices continues  Rate Limit jobs entities in the fleet-controller log, Applied work around to speed down the log flow ([FLEET_CLUSTER_ENQUEUE_DELAY environment variable ](https://github.com/rancher/fleet/pull/1071)to 120s) which is not in the v2.7.2.  Good thing is customer not noticed High CPU issue in v2.7.2. Created New Internal bug for the tracking purpose SURE-6279.

**Comment by moio** [comment]
Thanks @sadhugit, I will take a look at details there and turn it into a new GitHub issue if necessary!
